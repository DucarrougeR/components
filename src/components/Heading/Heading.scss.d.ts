export const lensHeading: string;
export const isDisplay1: string;
export const isDisplay2: string;
export const isDisplay3: string;
export const isHeading1: string;
export const isHeading2: string;
export const isHeading3: string;
export const isHeading4: string;
export const isHeading5: string;
export const isHeading6: string;
export const fontWeightLight: string;
export const fontWeightNormal: string;
export const fontWeightSemiBold: string;
export const fontWeightBold: string;
export const fontWeightExtraBold: string;
export const transformUpper: string;
export const transformLower: string;
export const transformCaps: string;
export const transformNone: string;
