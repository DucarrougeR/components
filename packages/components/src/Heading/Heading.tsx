import omit from 'lodash/omit'
import React, { FunctionComponent, Ref } from 'react'
import styled, { StyledComponent } from 'styled-components'
import {
  getTextTransform,
  ResponsiveFontSize,
  shouldTruncate,
  TextTransforms,
  textVariant,
  TextVariants,
} from '@looker/design-tokens'
import { Box, BoxProps } from '../Box'

export type HeadingLevels = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6'
export type HeadingTextTransforms = 'caps' | 'lower' | 'none' | 'upper'

const headingLevelSize = (as?: HeadingLevels) => {
  switch (as) {
    case 'h1':
      return 'xxlarge'
    case 'h2':
      return 'xlarge'
    case 'h3':
      return 'large'
    case 'h4':
      return 'medium'
    case 'h5':
      return 'small'
    case 'h6':
      return 'xsmall'
    default:
      return 'large'
  }
}

const headingLineHeight = (as?: HeadingLevels, size?: ResponsiveFontSize) => {
  if (size) return size

  switch (as) {
    case 'h1':
      return 'xxlarge'
    case 'h2':
      return 'xlarge'
    case 'h3':
      return 'large'
    case 'h4':
      return 'medium'
    case 'h5':
      return 'small'
    case 'h6':
      return 'xsmall'
    default:
      return 'large'
  }
}

export interface HeadingProps
  extends Omit<BoxProps<HTMLHeadingElement>, 'truncate'> {
  /** Heading level from h1-h6 */
  as?: HeadingLevels
  /** Text tranform  */
  textTransform?: TextTransforms
  /** Truncate heading text */
  truncate?: boolean
  /** Truncate heading at a specified number of lines (whole number) */
  truncateLines?: number
  /** Custom css class */
  className?: string
  /** Adjust style of text with more meaning by using an intent */
  variant?: TextVariants
}

type ComponentType = FunctionComponent<HeadingProps>
type StyledComponentType = StyledComponent<ComponentType, HeadingProps>

const InternalHeading: ComponentType = ({
  fontSize,
  fontWeight,
  lineHeight,
  as,
  ...props
}) => (
  <Box
    as={as || 'h2'}
    fontSize={fontSize || headingLevelSize(as)}
    lineHeight={lineHeight || headingLineHeight(as, fontSize)}
    fontWeight={fontWeight || 'normal'}
    {...omit(props, ['textTransform', 'truncate', 'truncateLines'])}
  >
    {props.children}
  </Box>
)

const HeadingFactory = React.forwardRef<StyledComponentType, HeadingProps>(
  (props: HeadingProps, ref: Ref<StyledComponentType>) => (
    <InternalHeading ref={ref} {...props} />
  )
)

export const Heading = styled<ComponentType>(HeadingFactory)`
  ${props => getTextTransform(props.textTransform)};
  ${props => shouldTruncate(props.truncate, props.truncateLines)};
  ${props => textVariant(props.theme, props.variant)};
`
