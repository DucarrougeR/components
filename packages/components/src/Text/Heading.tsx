import styled from 'styled-components'
import { ResponsiveValue } from 'styled-system'
import {
  CompatibleHTMLProps,
  FontSizes,
  textTransform,
  TextTransformProps,
} from 'looker-design-tokens'
import { TextBase, TextBaseProps } from './TextBase'
import { TextVariantProps, textVariant } from './text_variant'
import { TruncateProps, truncate } from './truncate'

type HeadingLevels = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6'

export interface HeadingProps
  extends TextBaseProps,
    TextTransformProps,
    TextVariantProps,
    TruncateProps,
    Omit<CompatibleHTMLProps<HTMLHeadingElement>, 'wrap'> {
  /** Heading level from h1-h6
   * @default: 'h2'
   */
  as?: HeadingLevels
}

const headingLevelFontSize = (props: HeadingProps) => {
  switch (props.as) {
    case 'h1':
      return 'xxlarge'
    case 'h3':
      return 'large'
    case 'h4':
      return 'medium'
    case 'h5':
      return 'small'
    case 'h6':
      return 'xsmall'
    case 'h2':
    default:
      return 'xlarge'
  }
}

const headingLineHeight = (props: HeadingProps): ResponsiveValue<FontSizes> =>
  props.fontSize ? props.fontSize : headingLevelFontSize(props)

export const Heading = styled(TextBase).attrs((props: HeadingProps) => ({
  fontSize: props.fontSize || headingLevelFontSize(props),
  lineHeight: props.lineHeight || headingLineHeight(props),
}))<HeadingProps>`
  ${textTransform}
  ${textVariant}
  ${truncate}
`

Heading.defaultProps = { as: 'h2', fontWeight: 'normal', variant: 'default' }
