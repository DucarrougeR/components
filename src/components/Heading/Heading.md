```js
<Heading className="cool" size="d1">Hello</Heading>
<Heading size="d2" level="2">Hello</Heading>
<Heading size="d3" level="3">Hello</Heading>
<Heading size="1" level="1">Hello</Heading>
<Heading size="2" level="2">Hello</Heading>
<Heading size="3" level="3">Hello</Heading>
<Heading size="4" level="4">Hello</Heading>
<Heading size="5" level="5">Hello</Heading>
<Heading level="6" size="">Hello</Heading>
```
