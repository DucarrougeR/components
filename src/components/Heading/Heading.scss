@import '../../styles/util';
@import '../../styles/variables';

.lens-heading {
  margin: 0;
  text-rendering: optimizeLegibility;
  font-family: $brand-font;
  font-weight: $default-font-weight;
}

.is-display-1 {
  font-size: font-size(d1);
  line-height: line-height(d1);
}

.is-display-2 {
  font-size: font-size(d2);
  line-height: line-height(d2);
}

.is-display-3 {
  font-size: font-size(d3);
  line-height: line-height(d3);
}

h1, .is-heading-1 {
  font-size: font-size(1);
  line-height: line-height(1);
}

h2, .is-heading-2 {
  font-size: font-size(2);
  line-height: line-height(2);
}

h3, .is-heading-3 {
  font-size: font-size(3);
  line-height: line-height(3);
}

h4, .is-heading-4 {
  font-size: font-size(4);
  line-height: line-height(4);
}

h5, .is-heading-5 {
  font-size: font-size(5);
  line-height: line-height(5);
}

h6, .is-heading-6 {
  font-size: font-size(6);
  line-height: line-height(6);
}

.font-weight-light {
  font-weight: 300;
}

.font-weight-normal {
  font-weight: 400;
}

.font-weight-semi-bold {
  font-weight: 600;
}

.font-weight-bold {
  font-weight: 700;
}

.font-weight-extra-bold {
  font-weight: 600;
}

.transform-upper {
  text-transform: uppercase;
}

.transform-lower {
  text-transform: lowercase;
}

.transform-caps {
  text-transform: capitalize;
}

.transform-none {
  text-transform: none;
}

