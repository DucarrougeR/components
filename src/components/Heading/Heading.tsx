import * as React from 'react'
import styled, { css } from '../../styled_components'
import { truncate } from '../../styles/typography'
import { lineHeights } from '../../styles/line_heights'
import { fontWeights } from '../../styles/font_weights'
import { RampSizes } from '../../styles/ramp_sizes'

export enum HeadingAlignments {
  Left = 'left',
  Center = 'center',
  Right = 'right'
}

export enum HeadingLevels {
  L1 = '1',
  L2 = '2',
  L3 = '3',
  L4 = '4',
  L5 = '5',
  L6 = '6'
}

export enum HeadingTextTransforms {
  Caps = 'caps',
  Lower = 'lower',
  None = 'none',
  Upper = 'upper'
}

export enum HeadingWeights {
  Bold = 'bold',
  ExtraBold = 'extraBold',
  Light = 'light',
  Normal = 'normal',
  SemiBold = 'semiBold'
}

export interface HeadingGeneratorProps {
  /** Headling tag level mapping for h1-h6 */
  level?: HeadingLevels
}

export interface HeadingProps extends HeadingGeneratorProps {
  /** Size mapping from type ramp */
  size?: RampSizes
  /** Font weight */
  weight?: HeadingWeights
  /** Text tranform  */
  transform?: HeadingTextTransforms
  /** Text align */
  align?: HeadingAlignments
  className?: string
  /** Truncate text on overflow */
  truncate?: boolean
}

const convertHeadingLevelToRampSize = (level: HeadingLevels | undefined) => {
  switch (level) {
    case HeadingLevels.L1:
      return RampSizes.One
    case HeadingLevels.L2:
      return RampSizes.Two
    case HeadingLevels.L3:
      return RampSizes.Three
    case HeadingLevels.L4:
      return RampSizes.Four
    case HeadingLevels.L5:
      return RampSizes.Five
    case HeadingLevels.L6:
      return RampSizes.Six
    default:
      return RampSizes.Three
  }
}

/**
 * Headings are used to help users understand  what a major section of an interface is about, for example the labeling
 * of a page or a title of a card component.
 */
const HeadingGenerator: React.SFC<HeadingProps> = ({ level, ...args }) => {
  // This prevents our props from being passed directly to the underlying h* tags, which ultimately
  // would cause some warnings. Ideally we would define the return type for this function, and
  // Typescript would warn us when passing props that are invalid.
  //
  // See https://reactjs.org/warnings/unknown-prop.html
  const props = Object.assign({}, args)
  delete props.align
  delete props.size
  delete props.transform
  delete props.truncate
  delete props.weight

  switch (level) {
    case HeadingLevels.L1:
      return <h1 {...props}>{props.children}</h1>
    case HeadingLevels.L2:
      return <h2 {...props}>{props.children}</h2>
    case HeadingLevels.L4:
      return <h4 {...props}>{props.children}</h4>
    case HeadingLevels.L5:
      return <h5 {...props}>{props.children}</h5>
    case HeadingLevels.L6:
      return <h6 {...props}>{props.children}</h6>
    case HeadingLevels.L3:
    default:
      return <h3 {...props}>{props.children}</h3>
  }
}

function textTransform(transform: HeadingTextTransforms | undefined) {
  switch (transform) {
    case HeadingTextTransforms.Upper:
      return css`
        text-transform: uppercase;
      `
    case HeadingTextTransforms.Lower:
      return css`
        text-transform: lowercase;
      `
    case HeadingTextTransforms.Caps:
      return css`
        text-transform: capitalize;
      `
    case HeadingTextTransforms.None:
    default:
      return css`
        text-transform: none;
      `
  }
}

function alignment(align: HeadingAlignments | undefined) {
  return css`
    text-align: ${align || HeadingAlignments.Left};
  `
}

export const Heading = styled<HeadingProps>(HeadingGenerator)`
  font-size: ${props =>
    props.theme.fontRamp[
      props.size || convertHeadingLevelToRampSize(props.level)
    ]};
  line-height: ${props =>
    lineHeights[props.size || convertHeadingLevelToRampSize(props.level)]};
  font-weight: ${props => fontWeights[props.weight || HeadingWeights.Normal]};
  ${props => textTransform(props.transform)}
  ${props => alignment(props.align)}
  ${props => truncate(props.truncate || false)}
`
