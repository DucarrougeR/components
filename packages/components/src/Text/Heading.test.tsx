import React from 'react'
import { createWithTheme } from 'looker-components-test-utils'
import { Heading } from './Heading'

test('A default Heading', () => {
  const component = createWithTheme(<Heading>ðŸ¥‘</Heading>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A <h1> Heading', () => {
  const component = createWithTheme(<Heading as="h1">ðŸ¥‘</Heading>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A <h1> Heading sized to <h2>', () => {
  const component = createWithTheme(
    <Heading as="h1" fontSize="xlarge">
      ðŸ¥‘
    </Heading>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A Heading to bold', () => {
  const component = createWithTheme(
    <Heading fontSize="large" fontWeight="bold">
      ðŸ¥‘
    </Heading>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A Heading transformed', () => {
  const component = createWithTheme(
    <Heading textTransform="capitalize">ðŸ¥‘</Heading>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A Heading with variant', () => {
  const component = createWithTheme(<Heading variant="subdued">ðŸ¥‘</Heading>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A Heading truncated', () => {
  const component = createWithTheme(<Heading truncate>ðŸ¥‘</Heading>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A Heading with multiline truncate', () => {
  const component = createWithTheme(<Heading truncateLines={2}>ðŸ¥‘</Heading>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
