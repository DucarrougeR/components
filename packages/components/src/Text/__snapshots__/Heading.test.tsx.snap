// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A <h1> Heading 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  font-weight: 400;
  line-height: 2.5rem;
  font-size: 1.5625rem;
}

<h1
  className="c0"
  fontSize="xxlarge"
  fontWeight="normal"
>
  ðŸ¥‘
</h1>
`;

exports[`A <h1> Heading sized to <h2> 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  font-size: 1.375rem;
  font-weight: 400;
  line-height: 2rem;
}

<h1
  className="c0"
  fontSize="xlarge"
  fontWeight="normal"
>
  ðŸ¥‘
</h1>
`;

exports[`A Heading to bold 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  font-size: 1.125rem;
  font-weight: 700;
  line-height: 1.75rem;
}

<h2
  className="c0"
  fontSize="large"
  fontWeight="bold"
>
  ðŸ¥‘
</h2>
`;

exports[`A Heading transformed 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  font-weight: 400;
  line-height: 2rem;
  font-size: 1.375rem;
  text-transform: capitalize;
}

<h2
  className="c0"
  fontSize="xlarge"
  fontWeight="normal"
>
  ðŸ¥‘
</h2>
`;

exports[`A Heading truncated 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  font-weight: 400;
  line-height: 2rem;
  font-size: 1.375rem;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

<h2
  className="c0"
  fontSize="xlarge"
  fontWeight="normal"
>
  ðŸ¥‘
</h2>
`;

exports[`A Heading with multiline truncate 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  font-weight: 400;
  line-height: 2rem;
  font-size: 1.375rem;
  overflow: hidden;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 2;
}

<h2
  className="c0"
  fontSize="xlarge"
  fontWeight="normal"
>
  ðŸ¥‘
</h2>
`;

exports[`A Heading with variant 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  font-weight: 400;
  line-height: 2rem;
  font-size: 1.375rem;
  color: #939BA5;
}

<h2
  className="c0"
  fontSize="xlarge"
  fontWeight="normal"
>
  ðŸ¥‘
</h2>
`;

exports[`A default Heading 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  font-weight: 400;
  line-height: 2rem;
  font-size: 1.375rem;
}

<h2
  className="c0"
  fontSize="xlarge"
  fontWeight="normal"
>
  ðŸ¥‘
</h2>
`;
